README.md
# ring-undertow

Clojure ring adapter to Undertow web server.

[![Clojars Project](https://img.shields.io/clojars/v/com.github.strojure/ring-undertow.svg)](https://clojars.org/com.github.strojure/ring-undertow)

[![cljdoc badge](https://cljdoc.org/badge/com.github.strojure/ring-undertow)](https://cljdoc.org/d/com.github.strojure/ring-undertow)
[![tests](https://github.com/strojure/ring-undertow/actions/workflows/tests.yml/badge.svg)](https://github.com/strojure/ring-undertow/actions/workflows/tests.yml)

## Motivation

- Performant implementation of the Ring handler concept.
- Lazy evaluation of request keys which are slow to compute but probably never
  used.
- Adopt features provided by underlying Undertow server, like sessions,
  websockets or multiple handlers on single server instance.

## Features

- The handler implementation is decoupled
  from [Undertow server API][github_undertow].

- Ring [request map][ring_requests]:

    - Does not contain deprecated keys.

    - `:headers` is a persistent map proxy interface over Undertow header map.
      Header names are case-insensitive. The proxy is converted to Clojure map
      on updates.

    - The request map itself is a [lazy map][github_zmap] with some values
      delayed.

    - Sessions work without additional middleware, only proper configuration of
      the Undertow server is required.

    - Does not contain rarely used keys like `:websocket?`. But this data can be
      got from request using functions from the `strojure.ring-undertow.request`
      namespace.

- Ring [response][ring_response] allows `HttpHandler` in the `:body`, this
  allows to easily initiate processing like websocket handshake.

## Usage

### Server

The library provides its own [functions][cljdoc_server] to start and stop
server. The [start][cljdoc_server_start] function enables ring handler adapter
(sync or async) for all functions specified as handlers in server configuration.

```clojure
(ns usage.server
  (:require [java-http-clj.core :as http]
            [strojure.ring-undertow.server :as server]
            [strojure.undertow.api.types :as types]))

(defn ring-handler
  "The ring handler function, both sync and async."
  ([request]
   {:body (str "Hello from " (:server-name request) " (sync)")})
  ([request respond raise]
   (future
     (try
       (respond {:body (str "Hello from " (:server-name request) " (async)")})
       (catch Throwable t
         (raise t))))))

(defn- run
  "Helper function to execute GET request, stop `server`, return response body."
  [server]
  (with-open [_ server]
    (let [port (-> server types/bean* :listenerInfo first :address :port)]
      (:body (http/send {:uri (str "http://localhost:" port "/")})))))

;; Synchronous ring handler
(run (server/start {:port 0
                    :handler ring-handler}))
;; or
(run (server/start {:port 0
                    :handler ring-handler
                    :ring-handler-type :sync}))

;; Asynchronous ring handler
(run (server/start {:port 0
                    :handler ring-handler
                    :ring-handler-type :async}))
```

### Handlers

Undertow server allow to use multiple handler functions in various contexts.
This library provides [functions to create server handlers][cljdoc_handler] from
Clojure functions. For server configuration documentation
see [Undertow server API][github_undertow].

```clojure
(ns usage.handlers
  (:require [java-http-clj.core :as http]
            [strojure.ring-undertow.handler :as ring.handler]
            [strojure.undertow.api.types :as types]
            [strojure.undertow.server :as server]))

(defn ring-handler
  "The ring handler function, both sync and async."
  ([request]
   {:body (str "Hello from " (:server-name request) " (sync)")})
  ([request respond raise]
   (future
     (try
       (respond {:body (str "Hello from " (:server-name request) " (async)")})
       (catch Throwable t
         (raise t))))))

(defn- run
  "Helper function to start server with `config`, execute HTTP request, stop
  server, return response body."
  [config]
  (with-open [server (server/start (assoc config :port 0))]
    (:body (http/send {:uri (str "http://localhost:"
                                 (-> server types/bean* :listenerInfo first :address :port)
                                 "/")}))))

;; ## Explicit invocation of handler function

;; Synchronous ring handler - explicitly
(run {:handler (ring.handler/sync ring-handler)})
;=> "Hello from localhost (sync)"

;; Asynchronous ring handler - explicitly
(run {:handler (ring.handler/async ring-handler)})
;=> "Hello from localhost (async)"

;; ## Using adapter configuration option

;; Synchronous ring handler - adapter in configuration
(run {:handler ring-handler, :handler-fn-adapter ring.handler/sync})
;=> "Hello from localhost (sync)"

;; Asynchronous ring handler - adapter in configuration
(run {:handler ring-handler, :handler-fn-adapter ring.handler/async})
;=> "Hello from localhost (async)"

;; ## Using declarative handler configuration
;;
;; NOTE: The `strojure.ring-undertow.handler` namespace need to be imported.

;; Synchronous ring handler - declarative
(run {:handler {:type ring.handler/ring :ring-handler ring-handler}})
;=> "Hello from localhost (sync)"

;; Asynchronous ring handler - declarative
(run {:handler {:type ::ring.handler/ring :ring-handler ring-handler
                :ring-handler-type :async}})
;=> "Hello from localhost (async)"

;; ## Setting adapter globally before server start

;; Synchronous ring handler - global assignment
(do (server/set-handler-fn-adapter ring.handler/sync)
    (run {:handler ring-handler}))
;=> "Hello from localhost (sync)"

;; Asynchronous ring handler - global assignment
(do (server/set-handler-fn-adapter ring.handler/async)
    (run {:handler ring-handler}))
;=> "Hello from localhost (async)"
```

### Websockets

Websocket server handler can be setup without involving ring handlers like
described in [Undertow server API][github_undertow].

But it is also possible to return websocket handler in response `:body` to
establish websocket connection.

```clojure
(ns usage.websockets
  (:require
    [strojure.ring-undertow.request :as request]
    [strojure.undertow.handler :as handler]
    [strojure.undertow.websocket.channel :as channel]))

(defn on-message
  "Websocket callback."
  [{:keys [channel text]}]
  (channel/send-text (str "Received: " text) channel nil))

(defn ring-handler
  "Ring handler initiating websocket connection."
  [request]
  (if (request/websocket? request)
    {:body (handler/websocket {:on-message on-message})}
    {:status 400}))
```

### Request utility

There are [some utility functions][cljdoc_request] to get information from ring
requests which is not presented as request map keys.

```clojure
(ns usage.request
  (:require [strojure.ring-undertow.request :as request]))

;; Check if sessions are enabled in Undertow server configuration.

(request/sessions-enabled? {})
;=> nil

;; Check if request is websocket upgrade request.

(request/websocket? {})
;=> nil

(request/websocket? {:headers {"upgrade" "unknown"}})
;=> false

(request/websocket? {:headers {"upgrade" "websocket"}})
;=> true
```

## Compatibility

There are another implementations of ring adapter for Undertow. This library
provides “adapters for adapters” for some of them:

- The [server.luminus-adapter][cljdoc_server_luminus] namespace for the
  [ring-undertow-adapter][github_luminus] in the Luminus framework.

## Benchmarks

- [request](doc/benchmark/ring_request.clj)
- [headers](doc/benchmark/ring_request_headers.clj)

---

[github_undertow]:
https://github.com/strojure/undertow

[ring_requests]:
https://github.com/ring-clojure/ring/wiki/Concepts#requests

[ring_response]:
https://github.com/ring-clojure/ring/wiki/Concepts#responses

[github_zmap]:
https://github.com/strojure/zmap

[github_luminus]:
https://github.com/luminus-framework/ring-undertow-adapter

[cljdoc_server]:
https://cljdoc.org/d/com.github.strojure/ring-undertow/CURRENT/api/strojure.ring-undertow.server

[cljdoc_server_start]:
https://cljdoc.org/d/com.github.strojure/ring-undertow/CURRENT/api/strojure.ring-undertow.server#start

[cljdoc_request]:
https://cljdoc.org/d/com.github.strojure/ring-undertow/CURRENT/api/strojure.ring-undertow.request

[cljdoc_handler]:
https://cljdoc.org/d/com.github.strojure/ring-undertow/CURRENT/api/strojure.ring-undertow.handler

[cljdoc_server_luminus]:
https://cljdoc.org/d/com.github.strojure/ring-undertow/CURRENT/api/strojure.ring-undertow.server.luminus-adapter

[handler/dispatch]:
https://cljdoc.org/d/com.github.strojure/undertow/CURRENT/api/strojure.undertow.handler#dispatch

[handler/session]:
https://cljdoc.org/d/com.github.strojure/undertow/CURRENT/api/strojure.undertow.handler#session


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## `1.2.1-110-SNAPSHOT`

Release date `UNRELEASED`



## `1.2.0-109`

Release date `2023-04-27`

- (deps): upgrade strojure/undertow "1.4.0-125"
- (feat): dispatch async handler on SameThreadExecutor

## `1.1.1-105`

Release date `2023-03-28`

- (deps): upgrade strojure/undertow "1.3.1-120"
  + (feat): generating CSP nonce changed from 18 to 16 bytes, URL encoded

## `1.1.0-102`

Release date `2023-03-22`

- (deps feat): upgrade strojure/undertow "1.3.0-116"
  - (feat session): use `HttpOnly` and `SameSite=Lax` in session cookie defaults

## `1.0.99`

Release date `2023-03-21`

- (deps): upgrade strojure/undertow "1.2.0-112"

## `1.0.96`

Release date `2023-03-09`

- (deps): Upgrade undertow "1.0.92", zmap "1.3.26".

## `1.0.92`

Release date `2023-03-09`

- (chore): Change license to Unlicense.

## `1.0.86`

Release date `2023-03-03`

- \[ impl ] - Handle nil :body - Add type hint to fix reflection warning.
- \[ test ] - Add tests for request helper functions.
- \[ test ] - Test async handler timeout.

## `1.0.81`

Release date `2023-03-02`

- \[ impl ] - Implement timeout for async ring handler.
- \[ impl ] - Add type hint to `server/start`.
- \[ impl ] - Handle nil :body.
- \[ session ] - Overwrite ring session data accordingly to spec.
- \[ session ] - Handle :recreate meta, unbind session on delete.
- \[ test ] - Add ring request tests.
- \[ test ] - Add ring response tests.
- \[ test ] - Add ring session tests.
- \[ test ] - Add server tests.
- \[ deps ] - Upgrade undertow 1.0.88, zmap 1.3.18.

## `1.0.65-beta10`

Release date `2023-01-23`

- Use `exchange.getRequestURI` for request `:uri`.
- Revert changes incompatible with Ring middlewares:
  > - Accept `HttpHandler` as response instead of handler in the `:body`.
  > - Accept `IBlockingDeref` as response for async execution.

## `1.0.58-beta9`

Release date `2023-01-19`

- ~~Use `exchange.getRequestPath` instead of `.getRequestURI` which can contains
  host.~~
- ~~Accept `HttpHandler` as response instead of handler in the `:body`.~~
- ~~Accept `IBlockingDeref` as response for async execution.~~
- Upgrade dependencies - strojure/undertow "1.0.72-rc2", zmap "1.2.11".

## `1.0.52-beta8`

Release date `2023-01-13`

- Upgrade dependencies - strojure/undertow "1.0.69-rc1", zmap "1.1.5".
- Change `toString` for HeaderMapProxy to use underlying `HeaderMap.toString`.

## `1.0.45-beta7`

Release date `2022-12-25`

- Implement configuration adapter for Luminus ring adapter.
- Upgrade undertow API to version `1.0.64-beta9`

## `1.0.41-beta6`

Release date `2022-12-24`

- Implement `server` namespace for convenient start/stop.
- Rename handler functions:
    - `handler/ring-sync` to `sync`;
    - `handler/ring-async` to `async`.
- Change declarative handler API for consistency with server configuration.

## `1.0.36-beta5`

Release date `2022-12-24`

- Upgrade undertow API to version `1.0.60-beta8`

## `1.0.33-beta4`

Release date `2022-12-23`

- Rename handler functions
    - `handler/sync-ring-handler` to `ring-sync`
    - `handler/async-ring-handler` to `ring-async`
- Upgrade undertow API to version `1.0.57-beta7`
- #1 Implement declarative configuration of ring handler

## `1.0.25-beta.3`

Release date `2022-12-22`

- Upgrade Undertow API to version `1.0.47-beta.6`.

## `1.0.17-beta.2`

Release date `2022-12-21`

- Rename `adapter` namespace to `handler`.
- Add various documentation.
- Upgrade Undertow API.

## `1.0.6-beta1`

Release date `2022-12-20`

- Initial implementation.


project.clj
(defproject com.github.strojure/ring-undertow "1.2.1-110-SNAPSHOT"
  :description "Clojure ring adapter to Undertow web server."
  :url "https://github.com/strojure/ring-undertow"
  :license {:name "The Unlicense" :url "https://unlicense.org"}

  :dependencies [;; Undertow server API.
                 [com.github.strojure/undertow "1.4.0-125"]
                 ;; Lazy map for ring request.
                 [com.github.strojure/zmap "1.3.26"]]

  :profiles {:provided {:dependencies [[org.clojure/clojure "1.11.1"]]}
             :dev,,,,, {:dependencies [;; Competitor libraries
                                       [luminus/ring-undertow-adapter "1.3.0"]
                                       [org.immutant/web "2.1.10"]
                                       ;; Testing HTTP requests
                                       [java-http-clj "0.4.3"]]
                        :source-paths ["doc"]}}

  :deploy-repositories [["clojars" {:url "https://clojars.org/repo" :sign-releases false}]])
